flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java

```diff
@@ -213,9 +213,9 @@ public static void setTokensFor(
 
         Collection<Token<? extends TokenIdentifier>> usrTok = currUsr.getTokens();
         for (Token<? extends TokenIdentifier> token : usrTok) {
-            final Text id = new Text(token.getIdentifier());
-            LOG.info("Adding user token " + id + " with " + token);
-            credentials.addToken(id, token);
+            final Text alias = new Text(token.getService());
+            LOG.info("Adding user token " + alias + " with " + token);
+            credentials.addToken(alias, token);
```

Comment:
Could you help me to understand why we need this and what the different between `token.getService()` and `token.getIdentifier()`. Thanks

Reply:
I found that the `identifier` is not the only uniq id of the token.
So when getting existed token from `UserGroupInformation.getCurrentUser().getTokens` and these token may be having the same `identifier`, this is will cause overwriting tokens by using  `credentials.addToken(identifier, token);`

Acutally, i think we should use `service` as the alias.

More detailed info has been replied in email to you.

Reply:
Looks like the token fetching for HBase is also relying on the service name instead of the identifier to add the token. I don't understand your argument, though: In what situation would the identifier not be unique. ...considering that you called the `identifier` not being the only unique id. Could you elaborate a bit more.

Reply:
According to digging into Hadoop source code, i found that `alias` is just as the key as internal `Credentials` token HashMap.

> In what situation would the identifier not be unique

In our production env, i found that `identifier` is the same in active and standby namenode delegation token. But i dont find any detailed Hadoop doc to support it.

Reply:
Are there different delegation tokens for the different namenodes Flink has to deal with? I would have had assumed that there's only a single token for Flink accessing the Yarn cluster. Could you point me to the Hadoop source code you're referring to.

Sorry if it's a bit tedious. I haven't worked with Yarn/Kerberos, yet

Reply:
Refer to https://issues.apache.org/jira/browse/HDFS-9276